
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>VcfQC &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-VCF.VcfQC">
<span id="vcfqc"></span><h1>VcfQC<a class="headerlink" href="#module-VCF.VcfQC" title="Permalink to this headline">¶</a></h1>
<p>VcfQC module.</p>
<p>This module is used to do a series of VCF QC tasks on a VCF file.</p>
<p>&#64;author: Ernesto Lowy</p>
<dl class="class">
<dt id="VCF.VcfQC.BcftoolsStats">
<em class="property">class </em><code class="descclassname">VCF.VcfQC.</code><code class="descname">BcftoolsStats</code><span class="sig-paren">(</span><em>filename=None</em>, <em>summary_numbers=None</em>, <em>ts_tv=None</em>, <em>ts_tv_1stalt=None</em>, <em>no_singleton_snps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.BcftoolsStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store the results of running BCFtools stats on a VCF file</p>
</dd></dl>

<dl class="class">
<dt id="VCF.VcfQC.CollectVCallingMetrics">
<em class="property">class </em><code class="descclassname">VCF.VcfQC.</code><code class="descname">CollectVCallingMetrics</code><span class="sig-paren">(</span><em>vc_detail_metrics_file</em>, <em>vc_summary_metrics_file</em>, <em>vc_detail_metrics=None</em>, <em>vc_summary_metrics=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.CollectVCallingMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store the results of running picard’s CollectVariantCallingMetrics on a VCF file</p>
</dd></dl>

<dl class="class">
<dt id="VCF.VcfQC.GTPconcordance">
<em class="property">class </em><code class="descclassname">VCF.VcfQC.</code><code class="descname">GTPconcordance</code><span class="sig-paren">(</span><em>summary_metrics_file</em>, <em>summary_metrics_snps=None</em>, <em>summary_metrics_indels=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.GTPconcordance" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store the results of running picard’s GenotypeConcordance on a VCF file</p>
</dd></dl>

<dl class="class">
<dt id="VCF.VcfQC.VcfQC">
<em class="property">class </em><code class="descclassname">VCF.VcfQC.</code><code class="descname">VcfQC</code><span class="sig-paren">(</span><em>vcf</em>, <em>bgzip_folder=None</em>, <em>bcftools_folder=None</em>, <em>bedtools_folder=None</em>, <em>picard_folder=None</em>, <em>r_folder=None</em>, <em>r_scripts=None</em>, <em>tabix_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to do quality assessment on a VCF file</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.calc_concordance" title="VCF.VcfQC.VcfQC.calc_concordance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_concordance</span></code></a>(truth_vcf,&nbsp;truth_sample,&nbsp;…)</td>
<td>Method to calculate the genotype concordance between VcfQC.vcf and Truth VCF.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.get_chros" title="VCF.VcfQC.VcfQC.get_chros"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_chros</span></code></a>([filter_str,&nbsp;chr_f,&nbsp;verbose])</td>
<td>Method to get a list of chromosomes present in a file</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.number_variants_in_region" title="VCF.VcfQC.VcfQC.number_variants_in_region"><code class="xref py py-obj docutils literal notranslate"><span class="pre">number_variants_in_region</span></code></a>(region,&nbsp;outprefix)</td>
<td>Method to get the number of variants in a particular region/s</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.plot_variant_density" title="VCF.VcfQC.VcfQC.plot_variant_density"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_variant_density</span></code></a>(length,&nbsp;genome,&nbsp;…)</td>
<td>Method used to plot the variant density along the genome for this VCF file</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.run_CollectVariantCallingMetrics" title="VCF.VcfQC.VcfQC.run_CollectVariantCallingMetrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_CollectVariantCallingMetrics</span></code></a>(outprefix,&nbsp;…)</td>
<td>Method to run Picard’s CollectVariantCallingMetrics on a VcfQC object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VcfQC.VcfQC.stats" title="VCF.VcfQC.VcfQC.stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stats</span></code></a>(outpath[,&nbsp;filter_str,&nbsp;region,&nbsp;…])</td>
<td>Run bcftools stats on the VCF file</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VcfQC.VcfQC.calc_concordance">
<code class="descname">calc_concordance</code><span class="sig-paren">(</span><em>truth_vcf</em>, <em>truth_sample</em>, <em>call_sample</em>, <em>outprefix</em>, <em>outdir=None</em>, <em>intervals=None</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.calc_concordance" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the genotype concordance between VcfQC.vcf and Truth VCF.
It will run Picard’s GenotypeConcordance</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>truth_vcf</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">The VCF containing the truth sample.</p>
</dd>
<dt><strong>truth_sample</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">The name of the truth sample within the truth VCF.</p>
</dd>
<dt><strong>call_sample</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">The name of the call sample within the call VCF.</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">String used as the prefix in the output file.</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">If provided, then put output files in this folder.</p>
</dd>
<dt><strong>intervals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">One or more interval list files that will be used to limit the
genotype concordance.</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Ff true, then print the command line used for running this program.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>GTPconcordance object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VcfQC.VcfQC.get_chros">
<code class="descname">get_chros</code><span class="sig-paren">(</span><em>filter_str=None</em>, <em>chr_f=None</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.get_chros" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to get a list of chromosomes present in a file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>filter_str</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">If defined, apply this filter string so bcftools view
apply it before fetching the chros.</p>
</dd>
<dt><strong>chr_f</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">Path to file with a list of chromosomes (one per line).
If provided, the chros in the file will be compared with the
chromosomes in self.vcf.</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Increase verbosity.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dict</strong></dt>
<dd><p class="first">Dict with a key named ‘in_vcf’ and whose values are the chros that are present in self.vcf.</p>
<dl class="last docutils">
<dt>If list_of_chros is defined, then it will also add 3 keys to the dict:</dt>
<dd><p class="first last">‘both’ whose values will be the chros present in self.vcf and in ‘chr_f’
‘in_A’ whose values will be the chros PRESENT in self.vcf and NOT in ‘chr_f’
‘in_B’ whose values will be the chros NOT present in self.vcf and PRESENT in ‘chr_f’.</p>
</dd>
</dl>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VcfQC.VcfQC.number_variants_in_region">
<code class="descname">number_variants_in_region</code><span class="sig-paren">(</span><em>region</em>, <em>outprefix</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.number_variants_in_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to get the number of variants in a particular region/s</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>region</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">String with path to the BED file containing the regions for
which the number will be calculated.</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Prefix for outfile.</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Increase verbosity.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong></dt>
<dd><p class="first last">File with the number of variants for each particular region.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VcfQC.VcfQC.plot_variant_density">
<code class="descname">plot_variant_density</code><span class="sig-paren">(</span><em>length</em>, <em>genome</em>, <em>outprefix</em>, <em>plot_params</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.plot_variant_density" title="Permalink to this definition">¶</a></dt>
<dd><p>Method used to plot the variant density along the genome for this VCF file</p>
<p>This method will calculate the variant density (SNPs, INDELs or both)
for windows of a certain length (in bp) defined by the user and will
plot this density along the chromosomes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>length</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">Length in bp of the genomic windows for which the number
of variants will be calculated.</p>
</dd>
<dt><strong>genome</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">File with genome sizes.</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Prefix for output file (png file with variant density plot).</p>
</dd>
<dt><strong>plot_params</strong> <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd><p class="first">Dictionary containing the graphical parameters used for plot_variants.R:
{</p>
<blockquote class="last">
<div><p>‘height’: int,
‘chromName_cex’: double
}.</p>
</div></blockquote>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong></dt>
<dd><p class="first last">Returns png file with the density plot</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VcfQC.VcfQC.run_CollectVariantCallingMetrics">
<code class="descname">run_CollectVariantCallingMetrics</code><span class="sig-paren">(</span><em>outprefix</em>, <em>truth_vcf</em>, <em>intervals=None</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.run_CollectVariantCallingMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to run Picard’s CollectVariantCallingMetrics on a VcfQC object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Prefix for outfiles: prefix.variant_calling_detail_metrics
and prefix.variant_calling_summary_metrics.</p>
</dd>
<dt><strong>truth_vcf</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Reference VCF file.</p>
</dd>
<dt><strong>intervals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">Target intervals to restrict analysis to.</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Increase verbosity.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>CollectVCallingMetrics object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VcfQC.VcfQC.stats">
<code class="descname">stats</code><span class="sig-paren">(</span><em>outpath</em>, <em>filter_str=None</em>, <em>region=None</em>, <em>region_file=None</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VcfQC.VcfQC.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Run bcftools stats on the VCF file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outpath</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">output path</p>
</dd>
<dt><strong>filter_str</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional.</span></dt>
<dd><p class="first last">Example:  PASS,.
Apply filters when calculating the stats.</p>
</dd>
<dt><strong>region</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">Example: chr20
Region used for calculating the stats.</p>
</dd>
<dt><strong>region_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">BED file with the regions that will be analyzed.</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>BcftoolsStats object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Auto-documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="VCF.html">VCF</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Auto-documentation</a><ul>
  <li><a href="VCF.html">VCF</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../_sources/code/VCF/VcfQC.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>