
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Whole Genome Sequencing BAM QC pipeline &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Auto-documentation" href="../code/index.html" />
    <link rel="prev" title="Workflows" href="index.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="whole-genome-sequencing-bam-qc-pipeline">
<h1>Whole Genome Sequencing BAM QC pipeline<a class="headerlink" href="#whole-genome-sequencing-bam-qc-pipeline" title="Permalink to this headline">¶</a></h1>
<ol class="arabic simple">
<li>Preparing the environment</li>
</ol>
<blockquote>
<div><p>Modify your $PYTHONPATH to include the required libraries:
export PYTHONPATH=${ehive_dir}/wrappers/python3/:$PYTHONPATH</p>
<p>Modify your $PERL5LIB to include the required libraries:
export PERL5LIB=${ehive_dir}/modules/:${igsr_analysis_dir}/:${PERL5LIB}</p>
<p>Modify your $PATH to include the location of the eHive scripts: export PATH=${ehive_dir}/scripts/:${PATH}</p>
<p>Install dependency:</p>
<ul class="simple">
<li>Clone repo by doing git clone <a class="reference external" href="https://github.com/igsr/igsr_analysis.git">https://github.com/igsr/igsr_analysis.git</a> in the desired folder</li>
<li>pip install ${igsr_analysis_dir}/dist/igsr_analysis-0.91.dev0.tar.gz</li>
<li>Modify $PYTHONPATH to add the folder where your pip installs the Python packages</li>
</ul>
<p>And you are ready to go!</p>
<p>Conventions used in this section:</p>
<p>${igsr_analysis_dir} is the folder where you have cloned <a class="reference external" href="https://github.com/igsr/igsr_analysis.git">https://github.com/igsr/igsr_analysis.git</a>
${ehive_dir} is the folder where you have cloned <a class="reference external" href="https://github.com/Ensembl/ensembl-hive.git">https://github.com/Ensembl/ensembl-hive.git</a></p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li>Databases</li>
</ol>
<blockquote>
<div><p>The pipeline uses two databases. They may be on different servers or the
same server.</p>
<p>2.1 The ReseqTrack database</p>
<blockquote>
<div><p>The pipeline queries a ReseqTrack database to find the VCF that will be
filtered by the pipeline. It will also add file metadata for the final
filtered VCF.</p>
<p>In order to create a ReseqTrack database use the following</p>
<p>commands:</p>
<dl class="docutils">
<dt>mysql -h &lt;hostname&gt; -P &lt;portnumber&gt; -u &lt;username&gt; -p???? -e “create database testreseqtrack” # where testreseqtrack</dt>
<dd><dl class="first last docutils">
<dt># is the name you want</dt>
<dd># to give to the ReseqTrack DB</dd>
</dl>
</dd>
</dl>
<p>mysql -h &lt;hostname&gt; -P &lt;portnumber&gt; -u &lt;username&gt; -p???? testreseqtrack &lt; $RESEQTRACK/sql/table.sql
mysql -h &lt;hostname&gt; -P &lt;portnumber&gt; -u &lt;username&gt; -p???? testreseqtrack &lt; $RESEQTRACK/sql/views.sql</p>
<p>Conventions used in this section:</p>
<p>$RESEQTRACK is the folder where you have cloned <a class="reference external" href="https://github.com/EMBL-EBI-GCA/reseqtrack.git">https://github.com/EMBL-EBI-GCA/reseqtrack.git</a></p>
</div></blockquote>
<p>2.2 The Hive database</p>
<blockquote>
<div>This is database is used by the Hive code to manage the pipeline and job
submission etc. The pipeline will be created automatically when you run
the init_pipeline.pl script.  Write access is needed to this database.</div></blockquote>
</div></blockquote>
<ol class="arabic simple" start="3">
<li>Initialise the pipeline</li>
</ol>
<blockquote>
<div><blockquote>
<div><p>The pipeline is initialised with the hive script init_pipeline.pl. Here is
an example of how to initialise a pipeline:</p>
<dl class="docutils">
<dt>init_pipeline.pl PyHive::PipeConfig::QC::RunBamQCsonWGS </dt>
<dd>-pipeline_url mysql://g1krw:$DB_PASS&#64;mysql-rs-1kg-prod:4175/hive_dbname -db testreseqtrack -pwd $DB_PASS -hive_force_init 1</dd>
</dl>
<p>The first argument is the the module that defines this pipeline.
Then -pipeline_url controls the Hive database connection details, in this
example:</p>
<blockquote>
<div>g1krw= username
$DB_PASS= password
mysql-rs-1kg-prod= hostname
4175= Port number
hive_dbname= Hive DB name</div></blockquote>
</div></blockquote>
<p>-db is the name of the Reseqtrack database name used in the section 2.1
-pwd is the ReseqTrack DB password</p>
<p>The rest of the options are documented in the PyHive::PipeConfig::QC::RunBamQCsonWGS
module file. You will probably want to override the defaults for many of
these options so take a look.</p>
</div></blockquote>
<ol class="arabic simple" start="4">
<li>Seeding the pipeline</li>
</ol>
<blockquote>
<div><p>In order to seed the pipeline with the VCF file that will be analyzed use the hive script seed_pipeline.pl: </p>
<dl class="docutils">
<dt>seed_pipeline.pl </dt>
<dd>-url mysql://g1krw:$DB_PASS&#64;mysql-rs-1kg-prod:4175/hive_dbname -logic_name find_files -input_id “{ ‘file’ =&gt; ‘/path/to/file/input_file.txt’ }”</dd>
</dl>
<p>Where -url controls the Hive database connection details and ‘/path/to/file/input_file.txt’
contains the filename of the VCF to be analyzed. This file must exist in the ReseqTrack database</p>
</div></blockquote>
<ol class="arabic simple" start="5">
<li>Sync the hive database</li>
</ol>
<blockquote>
<div><p>This should always be done before [re]starting a pipeline:</p>
<dl class="docutils">
<dt>Run e.g.:</dt>
<dd>beekeeper.pl -url mysql://g1krw:{password}&#64;mysql-g1k:4175/my_hive_db_name -sync</dd>
</dl>
<p>where -url is the details of your hive database.  Look at the output from
init_pipeline.pl to see what your url is.</p>
</div></blockquote>
<ol class="arabic simple" start="6">
<li>Run the pipeline</li>
</ol>
<blockquote>
<div><dl class="docutils">
<dt>Run e.g.:</dt>
<dd>beekeeper.pl -url mysql://g1krw:{password}&#64;mysql-g1k:4175/my_hive_db_name -loop &amp;</dd>
</dl>
<p>Note the ‘&amp;’ makes it run in the background.</p>
<p>Look at the pod for beekeeper.pl to see the various options.  E.g. you might
want to use the -hive_log_dir flag so that all output/error gets recorded in
files.</p>
<p>While the pipeline is running, you can check the ‘progress’ view of the hive
database to see the current status.  If a job has failed, check the msg
view.</p>
</div></blockquote>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Workflows</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Whole Genome Sequencing BAM QC pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../code/index.html">Auto-documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Workflows</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Workflows</a></li>
      <li>Next: <a href="../code/index.html" title="next chapter">Auto-documentation</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/workflows/wgs_bamqc_pipeline.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>