
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>GATK &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-VCF.VCFfilter.GATK">
<span id="gatk"></span><h1>GATK<a class="headerlink" href="#module-VCF.VCFfilter.GATK" title="Permalink to this headline">¶</a></h1>
<p>Created on 27 Feb 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFfilter.GATK.GATK">
<em class="property">class </em><code class="descclassname">VCF.VCFfilter.GATK.</code><code class="descname">GATK</code><span class="sig-paren">(</span><em>vcf</em>, <em>reference</em>, <em>caller='UG'</em>, <em>bgzip_folder=None</em>, <em>tabix_folder=None</em>, <em>gatk_folder=None</em>, <em>tmp_dir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to filter a VCF file using different GATK components</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFfilter.GATK.GATK.run_applyrecalibration" title="VCF.VCFfilter.GATK.GATK.run_applyrecalibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_applyrecalibration</span></code></a>(mode,&nbsp;recal_file,&nbsp;…)</td>
<td>Run GATK’s ApplyRecalibration on a VcfFilter object</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFfilter.GATK.GATK.run_variantrecalibrator" title="VCF.VCFfilter.GATK.GATK.run_variantrecalibrator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_variantrecalibrator</span></code></a>(resources,&nbsp;mode[,&nbsp;…])</td>
<td>Run GATK’s VariantRecalibrator on a VcfFilter object</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFfilter.GATK.GATK.run_applyrecalibration">
<code class="descname">run_applyrecalibration</code><span class="sig-paren">(</span><em>mode</em>, <em>recal_file</em>, <em>tranches_file</em>, <em>outprefix</em>, <em>ts_filter_level=99.0</em>, <em>num_threads=1</em>, <em>compress=True</em>, <em>verbose=None</em>, <em>log_file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK.run_applyrecalibration" title="Permalink to this definition">¶</a></dt>
<dd><p>Run GATK’s ApplyRecalibration on a VcfFilter object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>mode</strong> <span class="classifier-delimiter">:</span> <span class="classifier">{‘SNP’,’INDEL’}</span></dt>
<dd><p class="first last">Recalibration mode to employ</p>
</dd>
<dt><strong>recal_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">The input recal file used by ApplyRecalibration</p>
</dd>
<dt><strong>tranches_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">The input tranches file describing where to cut the data</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Prefix used for the output</p>
</dd>
<dt><strong>ts_filter_level</strong> <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd><p class="first last">The truth sensitivity level at which to start filtering. Default=99.0</p>
</dd>
<dt><strong>num_threads</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">Number of data threads to allocate to this analysis. Default=1</p>
</dd>
<dt><strong>compress</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Compress the recalibrated VCF. Default= True</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Increase verbosity</p>
</dd>
<dt><strong>log_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">Path to file that will used for logging the GATK stderr and stdout</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong></dt>
<dd><p class="first last">Path to filtered VCF file</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFfilter.GATK.GATK.run_variantrecalibrator">
<code class="descname">run_variantrecalibrator</code><span class="sig-paren">(</span><em>resources</em>, <em>mode</em>, <em>max_gaussians=None</em>, <em>intervals=None</em>, <em>annotations=None</em>, <em>tranches=None</em>, <em>outprefix='recalibrate'</em>, <em>verbose=None</em>, <em>log_file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK.run_variantrecalibrator" title="Permalink to this definition">¶</a></dt>
<dd><p>Run GATK’s VariantRecalibrator on a VcfFilter object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>resources</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">JSON file with resources to add using the -resources option, Required</p>
</dd>
<dt><strong>mode</strong> <span class="classifier-delimiter">:</span> <span class="classifier">{‘SNP’,’INDEL’}</span></dt>
<dd><p class="first last">Recalibration mode to employ</p>
</dd>
<dt><strong>intervals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">chr1:1-1000, optional</span></dt>
<dd><p class="first last">One or more genomic intervals over which to operate</p>
</dd>
<dt><strong>max_gaussians</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">Max number of Gaussians for the positive model</p>
</dd>
<dt><strong>annotations</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list, optional</span></dt>
<dd><p class="first last">List of annotations to be used. Default=[‘DP’, ‘QD’, ‘FS’, ‘SOR’,
‘MQ’, ‘MQRankSum’, ‘ReadPosRankSum’, ‘InbreedingCoeff’]</p>
</dd>
<dt><strong>tranches</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list, optional</span></dt>
<dd><p class="first last">Each element in the list will correspond to the  levels of truth
sensitivity at which to slice the data. (in percent, that is 1.0
for 1 percent). Default=[100.0,99.9,99.0,90.0]</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">out prefix used for -recalFile, -tranchesFile, -rscriptFile.
Default= recalibrate</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Increase verbosity</p>
</dd>
<dt><strong>log_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">Path to file that will used for logging the GATK stderr and stdout</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dict</strong></dt>
<dd><p class="first last">Dictionary with location of tranches and recal files</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows/index.html">Workflows</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">Auto-documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../VCF.html">VCF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../BEDTools/BEDTools.html">BEDTools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../BamQC/BamQC.html">BamQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VariantCalling/VariantCalling.html">VariantCalling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Utils/Utils.html">Utils</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Auto-documentation</a><ul>
  <li><a href="../VCF.html">VCF</a><ul>
  <li><a href="VCFfilter.html">VCFfilter</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../../_sources/code/VCF/VCFfilter/GATK.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>