
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Beagle &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-VCF.VCFIntegration.Beagle">
<span id="beagle"></span><h1>Beagle<a class="headerlink" href="#module-VCF.VCFIntegration.Beagle" title="Permalink to this headline">¶</a></h1>
<p>Created on 21 Jul 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFIntegration.Beagle.Beagle">
<em class="property">class </em><code class="descclassname">VCF.VCFIntegration.Beagle.</code><code class="descname">Beagle</code><span class="sig-paren">(</span><em>vcf</em>, <em>beagle_folder=None</em>, <em>beagle_jar=None</em>, <em>makeBGLCHUNKS_folder=None</em>, <em>prepareGenFromBeagle4_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to operate on a VCF file and run Beagle and other Beagle-related operations on it</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks" title="VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_beagle_chunks</span></code></a>(window,&nbsp;overlap,&nbsp;outfile)</td>
<td>Method to generate the chromosome chunks for Beagle see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4" title="VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prepare_Gen_From_Beagle4</span></code></a>(prefix_in,&nbsp;outprefix)</td>
<td>Method that uses prepareGenFromBeagle4 in order to convert the different Beagle chunks generated by ‘self.make_beagle_chunks’ into a single concatenated output that can be used  with SHAPEIT.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.run_beagle" title="VCF.VCFIntegration.Beagle.Beagle.run_beagle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_beagle</span></code></a>(outprefix[,&nbsp;outdir,&nbsp;region,&nbsp;…])</td>
<td>Method that wraps Beagle (see <a class="reference external" href="https://faculty.washington.edu/browning/beagle/beagle.html">https://faculty.washington.edu/browning/beagle/beagle.html</a>) and will be used to call genotypes on a VCF file containing GT likelihoods</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks">
<code class="descname">make_beagle_chunks</code><span class="sig-paren">(</span><em>window</em>, <em>overlap</em>, <em>outfile</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to generate the chromosome chunks for Beagle
see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>window</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">The chunk size (–window) in number of variant sites</p>
</dd>
<dt><strong>overlap</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">The overlap size (–overlap) in number of variant sites</p>
</dd>
<dt><strong>outfile</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">Output file name. i.e. ‘chunk.coordinates’</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">If true, then print the command line used for running this tool.Default=True</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong></dt>
<dd><p class="first last">Path to file with the coordinates of the chunk</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4">
<code class="descname">prepare_Gen_From_Beagle4</code><span class="sig-paren">(</span><em>prefix_in</em>, <em>outprefix</em>, <em>threshold=0.995</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that uses prepareGenFromBeagle4 in order to convert the different Beagle chunks
generated by ‘self.make_beagle_chunks’ into a single concatenated output that can be used 
with SHAPEIT.
see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>prefix_in</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">prefix used in the output of the different Beagle chunks after running  method ‘self.run_beagle’.
i.e. output.beagle4.22.*.</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Prefix used for output files. i.e. If prefix ‘input.shapeit.chr22’ is used. Then it will generate the following files:
input.shapeit.chr22.gen.gz
input.shapeit.chr22.gen.sample
input.shapeit.chr22.hap.gz
input.shapeit.chr22.hap.sample</p>
</dd>
<dt><strong>threshold</strong> <span class="classifier-delimiter">:</span> <span class="classifier">float, optional</span></dt>
<dd><p class="first last">Threshold meaning that all genotypes with a posterior above 0.995 are directly fixed and will only need phasing in the SHAPEIT step.
Default: 0.995</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running this tool.Default=False</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dict</strong></dt>
<dd><p class="first last">A dict with the path to the 4 output files (<em>.gen.</em> and <em>.hap.</em>) that can be used with SHAPEIT</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.run_beagle">
<code class="descname">run_beagle</code><span class="sig-paren">(</span><em>outprefix</em>, <em>outdir=None</em>, <em>region=None</em>, <em>verbose=False</em>, <em>correct=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.run_beagle" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that wraps Beagle (see <a class="reference external" href="https://faculty.washington.edu/browning/beagle/beagle.html">https://faculty.washington.edu/browning/beagle/beagle.html</a>)
and will be used to call genotypes on a VCF file containing GT likelihoods</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outprefix: str</strong></dt>
<dd><p class="first last">Prefix used for output file</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">outdir for output files</p>
</dd>
<dt><strong>region</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">chr or chr interval that will be analyzed. i.e. chr20 or chr20:10000000-11000000</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running Beagle</p>
</dd>
<dt><strong>correct</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Note: that it seems there is an incompatibility between zlib libraries used in Beagle4 and in BOOST on some platforms.
This involves either the last line of the file being skipped or a segfault. If correct=True, then this function will fix this issue
by recompressing the Beagle4 output files. Default=False</p>
</dd>
<dt><strong>window</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">number of markers to include in each sliding
window. Default: 50000</p>
</dd>
<dt><strong>overlap</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">specifies the number of markers of overlap between sliding
windows. Default: 3000</p>
</dd>
<dt><strong>niterations</strong> <span class="classifier-delimiter">:</span> <span class="classifier">unt, optional</span></dt>
<dd><p class="first last">specifies the number of phasing iterations. Default:
niterations=5</p>
</dd>
<dt><strong>nthreads</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">number of threads. If not specified then the nthreads parameter 
will be set equal to the number of CPU cores on the host machine</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong></dt>
<dd><p class="first last">Compressed VCF file with the genotype calls</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../workflows/index.html">Workflows</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">Auto-documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../VCF.html">VCF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../BEDTools/BEDTools.html">BEDTools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../BamQC/BamQC.html">BamQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VariantCalling/VariantCalling.html">VariantCalling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Utils/Utils.html">Utils</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Auto-documentation</a><ul>
  <li><a href="../VCF.html">VCF</a><ul>
  <li><a href="VCFIntegration.html">VCFIntegration</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../../_sources/code/VCF/VCFIntegration/Beagle.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>