Benchmark with GIAB
===================

This workflow is used in our project to benchmark the sites identified in our call sets for sample NA12878 with the sites identified by the Genome in a Bottle consortium [https://github.com/genome-in-a-bottle]. 
This consortium was created to produce reference data sets that can be used for genome comparison and benchmarking and the high-confidence sites identified by this consortium for sample NA12878 are the result 
of an effort to integrate the data generated by 13 different sequencing technologies and analysis methods.

Dependencies
------------

* Nextflow
 This pipeline uses a workflow management system named Nextflow. This software can be downloaded from:

 https://www.nextflow.io/

* Tabix and bgzip
 Tabix and bgzip are part of the HTSlib project, which can be downloaded from:

 https://github.com/samtools/htslib

* BCFTools
 Downloadable from:

 http://www.htslib.org/download/

* BEDTools
 Downloadable from:

 https://github.com/arq5x/bedtools2/releases/

* vcf-validator
 This tool can be obtained from:

 https://github.com/EBIvariation/vcf-validator

* IGSR-analysis code base
 The scripts needed to run this workflow can be downloaded by cloning the IGSR-analysis github repo from:

 https://github.com/igsr/igsr_analysis.git

How to run the pipeline
-----------------------

