
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>VCF.VCFIntegration package &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="VCF package" href="VCF.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="vcf-vcfintegration-package">
<h1>VCF.VCFIntegration package<a class="headerlink" href="#vcf-vcfintegration-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-VCF.VCFIntegration.Beagle">
<span id="vcf-vcfintegration-beagle-module"></span><h2>VCF.VCFIntegration.Beagle module<a class="headerlink" href="#module-VCF.VCFIntegration.Beagle" title="Permalink to this headline">¶</a></h2>
<p>Created on 21 Jul 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFIntegration.Beagle.Beagle">
<em class="property">class </em><code class="descclassname">VCF.VCFIntegration.Beagle.</code><code class="descname">Beagle</code><span class="sig-paren">(</span><em>vcf</em>, <em>beagle_folder=None</em>, <em>beagle_jar=None</em>, <em>makeBGLCHUNKS_folder=None</em>, <em>prepareGenFromBeagle4_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to operate on a VCF file and run Beagle and other Beagle-related operations on it</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks" title="VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_beagle_chunks</span></code></a>(window,&nbsp;overlap,&nbsp;outfile)</td>
<td>Method to define chromosome chunks for Beagle see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4" title="VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prepare_Gen_From_Beagle4</span></code></a>(prefix_in,&nbsp;outprefix)</td>
<td>Method that uses prepareGenFromBeagle4 in order to convert the different Beagle chunks generated by ‘self.make_beagle_chunks’ into a single concatenated output that can be used  with SHAPEIT.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Beagle.Beagle.run_beagle" title="VCF.VCFIntegration.Beagle.Beagle.run_beagle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_beagle</span></code></a>(outprefix[,&nbsp;outdir,&nbsp;region,&nbsp;…])</td>
<td>Method that wraps Beagle (see <a class="reference external" href="https://faculty.washington.edu/browning/beagle/beagle.html">https://faculty.washington.edu/browning/beagle/beagle.html</a>) and will be used to call genotypes on a VCF file containing GT likelihoods</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks">
<code class="descname">make_beagle_chunks</code><span class="sig-paren">(</span><em>window</em>, <em>overlap</em>, <em>outfile</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.make_beagle_chunks" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to define chromosome chunks for Beagle
see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>window: int, required</strong></dt>
<dd><p class="first last">The chunk size (–window) in number of variant sites</p>
</dd>
<dt><strong>overlap: int, required</strong></dt>
<dd><p class="first last">The overlap size (–overlap) in number of variant sites</p>
</dd>
<dt><strong>outfile: str, required</strong></dt>
<dd><p class="first last">Name of output file. i.e. ‘chunk.coordinates’</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running this tool.Default=True</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Path to file with the coordinates of the chunk</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4">
<code class="descname">prepare_Gen_From_Beagle4</code><span class="sig-paren">(</span><em>prefix_in</em>, <em>outprefix</em>, <em>threshold=0.995</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.prepare_Gen_From_Beagle4" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that uses prepareGenFromBeagle4 in order to convert the different Beagle chunks
generated by ‘self.make_beagle_chunks’ into a single concatenated output that can be used 
with SHAPEIT.
see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gettingstarted</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>prefix_in: str, required</strong></dt>
<dd><p class="first last">prefix used in the output of the different Beagle chunks after running  method ‘self.run_beagle’.
i.e. output.beagle4.22.*.</p>
</dd>
<dt><strong>outprefix: str, required</strong></dt>
<dd><p class="first last">Prefix used for output files. i.e. If prefix ‘input.shapeit.chr22’ is used. Then it will generate the following files:
input.shapeit.chr22.gen.gz
input.shapeit.chr22.gen.sample
input.shapeit.chr22.hap.gz
input.shapeit.chr22.hap.sample</p>
</dd>
<dt><strong>threshold: float, optional</strong></dt>
<dd><p class="first last">Threshold meaning that all genotypes with a posterior above 0.995 are directly fixed and will only need phasing in the SHAPEIT step.
Default: 0.995</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running this tool.Default=False</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>A dict with the path to the 4 output files (*.gen.* and *.hap.*) that can be used with SHAPEIT</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Beagle.Beagle.run_beagle">
<code class="descname">run_beagle</code><span class="sig-paren">(</span><em>outprefix</em>, <em>outdir=None</em>, <em>region=None</em>, <em>verbose=False</em>, <em>correct=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Beagle.Beagle.run_beagle" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that wraps Beagle (see <a class="reference external" href="https://faculty.washington.edu/browning/beagle/beagle.html">https://faculty.washington.edu/browning/beagle/beagle.html</a>)
and will be used to call genotypes on a VCF file containing GT likelihoods</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outprefix: str, required</strong></dt>
<dd><p class="first last">Prefix used for output file</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">outdir for output files</p>
</dd>
<dt><strong>region</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">chr or chr interval that will be analyzed. i.e. chr20 or chr20:10000000-11000000</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running Beagle</p>
</dd>
<dt><strong>correct</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">Note: that it seems there is an incompatibility between zlib libraries used in Beagle4 and in BOOST on some platforms.
This involves either the last line of the file being skipped or a segfault. If correct=True, then this function will fix this issue
by recompressing the Beagle4 output files. Default=False</p>
</dd>
<dt><strong>window: int, optional</strong></dt>
<dd><p class="first last">number of markers to include in each sliding
window. Default: 50000</p>
</dd>
<dt><strong>overlap: int, optional</strong></dt>
<dd><p class="first last">specifies the number of markers of overlap between sliding
windows. Default: 3000</p>
</dd>
<dt><strong>niterations: unt, optional</strong></dt>
<dd><p class="first last">specifies the number of phasing iterations. Default:
niterations=5</p>
</dd>
<dt><strong>nthreads</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">number of threads. If not specified then the nthreads parameter 
will be set equal to the number of CPU cores on the host machine</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Compressed VCF file with the genotype calls</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-VCF.VCFIntegration.SNPTools">
<span id="vcf-vcfintegration-snptools-module"></span><h2>VCF.VCFIntegration.SNPTools module<a class="headerlink" href="#module-VCF.VCFIntegration.SNPTools" title="Permalink to this headline">¶</a></h2>
<p>Created on 21 Jul 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFIntegration.SNPTools.SNPTools">
<em class="property">class </em><code class="descclassname">VCF.VCFIntegration.SNPTools.</code><code class="descname">SNPTools</code><span class="sig-paren">(</span><em>vcf</em>, <em>snptools_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.SNPTools.SNPTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to operate on a VCF file at the population level and perform
different SNPTools analysis on it</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_bamodel" title="VCF.VCFIntegration.SNPTools.SNPTools.run_bamodel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_bamodel</span></code></a>(sample,&nbsp;bamfiles[,&nbsp;outdir,&nbsp;verbose])</td>
<td>Method that wraps SNPTools’ bamodel on a VCF containing only Biallelic SNPs See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_poprob" title="VCF.VCFIntegration.SNPTools.SNPTools.run_poprob"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_poprob</span></code></a>(outprefix,&nbsp;rawlist[,&nbsp;outdir,&nbsp;verbose])</td>
<td>Method that wraps SNPTools’ poprob on a VCF containing only Biallelic SNPs See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_prob2vcf" title="VCF.VCFIntegration.SNPTools.SNPTools.run_prob2vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_prob2vcf</span></code></a>(probf,&nbsp;outprefix,&nbsp;chro[,&nbsp;…])</td>
<td>Method that wraps SNPTools’ prob2vcf on a VCF containing only Biallelic SNPs See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFIntegration.SNPTools.SNPTools.run_bamodel">
<code class="descname">run_bamodel</code><span class="sig-paren">(</span><em>sample</em>, <em>bamfiles</em>, <em>outdir=None</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_bamodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that wraps SNPTools’ bamodel on a VCF containing only Biallelic SNPs
See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>sample: str, required</strong></dt>
<dd><p class="first last">sample to analyze</p>
</dd>
<dt><strong>bamfiles</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, required</span></dt>
<dd><p class="first last">file containing the BAM path/s (one per line) for sample</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">outdir for output files</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running SNPTools</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Returns a *.raw file</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.SNPTools.SNPTools.run_poprob">
<code class="descname">run_poprob</code><span class="sig-paren">(</span><em>outprefix</em>, <em>rawlist</em>, <em>outdir=None</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_poprob" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that wraps SNPTools’ poprob on a VCF containing only Biallelic SNPs
See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outprefix: str, required</strong></dt>
<dd><p class="first last">prefix for <a href="#id1"><span class="problematic" id="id2">*</span></a>.prob file</p>
</dd>
<dt><strong>rawlist: str, required</strong></dt>
<dd><p class="first last">file containing the paths to the <a href="#id3"><span class="problematic" id="id4">*</span></a>.raw files generated by
‘run_snptools_bamodel’</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">outdir for output files</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running SNPTools</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Returns a *.prob file</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.SNPTools.SNPTools.run_prob2vcf">
<code class="descname">run_prob2vcf</code><span class="sig-paren">(</span><em>probf</em>, <em>outprefix</em>, <em>chro</em>, <em>outdir=None</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.SNPTools.SNPTools.run_prob2vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Method that wraps SNPTools’ prob2vcf on a VCF containing only Biallelic SNPs
See <a class="reference external" href="https://www.hgsc.bcm.edu/software/snptools">https://www.hgsc.bcm.edu/software/snptools</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>probf: str, required</strong></dt>
<dd><p class="first last"><a href="#id5"><span class="problematic" id="id6">*</span></a>.prob file generated by ‘run_snptools_poprob’</p>
</dd>
<dt><strong>outprefix: str, required</strong></dt>
<dd><p class="first last">Prefix used for output file</p>
</dd>
<dt><strong>chro: str, required</strong></dt>
<dd><p class="first last">chromosome for which the vcf will be generated</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">outdir for output files</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running SNPTools</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Compressed VCF file with the population genotype likelihoods</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-VCF.VCFIntegration.Shapeit">
<span id="vcf-vcfintegration-shapeit-module"></span><h2>VCF.VCFIntegration.Shapeit module<a class="headerlink" href="#module-VCF.VCFIntegration.Shapeit" title="Permalink to this headline">¶</a></h2>
<p>Created on 21 Jul 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFIntegration.Shapeit.Shapeit">
<em class="property">class </em><code class="descclassname">VCF.VCFIntegration.Shapeit.</code><code class="descname">Shapeit</code><span class="sig-paren">(</span><em>bgzip_folder=None</em>, <em>shapeit_folder=None</em>, <em>ligateHAPLOTYPES_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Shapeit.Shapeit" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to run SHAPEIT</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Shapeit.Shapeit.convert2vcf" title="VCF.VCFIntegration.Shapeit.Shapeit.convert2vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">convert2vcf</span></code></a>(input_prefix,&nbsp;output_prefix[,&nbsp;…])</td>
<td>Function to use SHAPEIT’s -convert in order to convert the <a href="#id7"><span class="problematic" id="id8">*</span></a>.haps.gz &amp; <a href="#id9"><span class="problematic" id="id10">*</span></a>.haps.sample files into VCF</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFIntegration.Shapeit.Shapeit.ligate_shapeitchunks" title="VCF.VCFIntegration.Shapeit.Shapeit.ligate_shapeitchunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ligate_shapeitchunks</span></code></a>(vcf_f,&nbsp;…[,&nbsp;verbose])</td>
<td>Run ligateHAPLOTYPES to ligate together all haplotype chunks produced by SHAPEIT</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFIntegration.Shapeit.Shapeit.run_shapeit" title="VCF.VCFIntegration.Shapeit.Shapeit.run_shapeit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_shapeit</span></code></a>(output_prefix[,&nbsp;input_gen,&nbsp;…])</td>
<td>Run Shapeit</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFIntegration.Shapeit.Shapeit.convert2vcf">
<code class="descname">convert2vcf</code><span class="sig-paren">(</span><em>input_prefix</em>, <em>output_prefix</em>, <em>compress=False</em>, <em>verbose=False</em>, <em>logfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Shapeit.Shapeit.convert2vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to use SHAPEIT’s -convert in order to convert the <a href="#id11"><span class="problematic" id="id12">*</span></a>.haps.gz &amp; <a href="#id13"><span class="problematic" id="id14">*</span></a>.haps.sample files into VCF</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>input_prefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">Prefix for the files in HAPS/SAMPLE format</p>
</dd>
<dt><strong>output_prefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">String with the output prefix for the VCF file</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running this program</p>
</dd>
<dt><strong>logfile</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">Path for log file</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>A VCF file</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Shapeit.Shapeit.ligate_shapeitchunks">
<code class="descname">ligate_shapeitchunks</code><span class="sig-paren">(</span><em>vcf_f</em>, <em>scaffolded_samples</em>, <em>chunk_str</em>, <em>output_prefix</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Shapeit.Shapeit.ligate_shapeitchunks" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>Run ligateHAPLOTYPES to ligate together all haplotype chunks produced by SHAPEIT
(see <a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#haplegsample">https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#haplegsample</a>)</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>vcf_f</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><blockquote class="first">
<div><p>VCF with the Genotype likelihoods</p>
</div></blockquote>
<dl class="last docutils">
<dt>scaffolded_samples <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">File with the list of samples (separated by ‘</p>
</dd>
</dl>
</dd>
<dt><strong>‘) that have been scaffolded</strong></dt>
<dd><dl class="first last docutils">
<dt>chunk_str <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">String with the paths to the different files generated by SHAPEIT for the different 
chromosome chunks (i.e. ‘s2.chunk1.hap.gz s2.chunk1.hap.gz s2.chunk1.hap.gz’)</p>
</dd>
<dt>output_prefix <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">String with the output prefixes (i.e. ‘output.shapeit.22.ligated.haps.gz output.shapeit.22.ligated.haps.sample’)</p>
</dd>
</dl>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>A dict with the path to the 2 output files (*.haps.gz and *.haps.sample)</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFIntegration.Shapeit.Shapeit.run_shapeit">
<code class="descname">run_shapeit</code><span class="sig-paren">(</span><em>output_prefix</em>, <em>input_gen=None</em>, <em>input_init=None</em>, <em>input_scaffold=None</em>, <em>input_bed=None</em>, <em>duohmm=False</em>, <em>input_map=None</em>, <em>verbose=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFIntegration.Shapeit.Shapeit.run_shapeit" title="Permalink to this definition">¶</a></dt>
<dd><p>Run Shapeit</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>input_gen</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">specifies the genotype/GL input data that you obtain from Beagle4, i.e. ‘input.shapeit.20.gen.gz input.shapeit.20.gen.sample’</p>
</dd>
<dt><strong>input_init</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">specifies the haplotypes that you obtain from Beagle4, i.e. ‘input.shapeit.20.hap.gz input.shapeit.20.hap.sample’</p>
</dd>
<dt><strong>input_scaffold</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">SNP-array derived haplotype scaffold used by SHAPEIT. It has to be in Impute2 format. i.e. ‘scaffold.haps.gz scaffold.haps.sample’</p>
</dd>
<dt><strong>input_bed</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">Unphased genotypes in Plink Binary BED/BIM/FAM format. i.e. ‘file.bed file.bim file.fam’</p>
</dd>
<dt><strong>duohmm</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, Optional</span></dt>
<dd><p class="first last">If true, then activate the –duohmm option. Default: False</p>
</dd>
<dt><strong>output_prefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">Prefix used for the 2 output files estimated by SHAPEIT, i.e. ‘output.shapeit.20.haps.gz output.shapeit.20.haps.sample’</p>
</dd>
<dt><strong>input_map</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">Path with the genetic map</p>
</dd>
<dt><strong>i_from</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, Optional</span></dt>
<dd><p class="first last">specify the region to be phased</p>
</dd>
<dt><strong>i_to</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, Optional</span></dt>
<dd><p class="first last">specify the region to be phased</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first last">if true, then print the command line used for running this program</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>A dict with the path to the 2 output files (*.haps.gz and *.haps.sample) that can be used with SHAPEIT</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-VCF.VCFIntegration">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-VCF.VCFIntegration" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="VCF.html">VCF package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="VCF.html#subpackages">Subpackages</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfNormalize">VCF.VcfNormalize module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfQC">VCF.VcfQC module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfUtils">VCF.VcfUtils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage/installation.html">Installation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="VCF.html">VCF package</a><ul>
      <li>Previous: <a href="VCF.html" title="previous chapter">VCF package</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/VCF.VCFIntegration.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>