
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>VCF.VCFfilter package &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="vcf-vcffilter-package">
<h1>VCF.VCFfilter package<a class="headerlink" href="#vcf-vcffilter-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-VCF.VCFfilter.BCFTools">
<span id="vcf-vcffilter-bcftools-module"></span><h2>VCF.VCFfilter.BCFTools module<a class="headerlink" href="#module-VCF.VCFfilter.BCFTools" title="Permalink to this headline">¶</a></h2>
<p>Created on 27 Feb 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFfilter.BCFTools.BCFTools">
<em class="property">class </em><code class="descclassname">VCF.VCFfilter.BCFTools.</code><code class="descname">BCFTools</code><span class="sig-paren">(</span><em>vcf</em>, <em>bcftools_folder=None</em>, <em>tabix_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.BCFTools.BCFTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to perform different filtering/selecting operations using BCFTools
<a class="reference external" href="https://samtools.github.io/bcftools/bcftools.html">https://samtools.github.io/bcftools/bcftools.html</a></p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFfilter.BCFTools.BCFTools.filter" title="VCF.VCFfilter.BCFTools.BCFTools.filter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter</span></code></a>(name,&nbsp;expression[,&nbsp;verbose])</td>
<td>Run bcftools filter on a Vcf file</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFfilter.BCFTools.BCFTools.filter_by_variant_type" title="VCF.VCFfilter.BCFTools.BCFTools.filter_by_variant_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_by_variant_type</span></code></a>(outprefix[,&nbsp;v_type,&nbsp;…])</td>
<td>Method to filter a VCF file by variant type. For example, to extract only the SNPs</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFfilter.BCFTools.BCFTools.select_variants" title="VCF.VCFfilter.BCFTools.BCFTools.select_variants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">select_variants</span></code></a>(outprefix[,&nbsp;uncalled,&nbsp;…])</td>
<td>Run bcftools view to select only the variants (exclude the 0|0 genotypes)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFfilter.BCFTools.BCFTools.subset_vcf" title="VCF.VCFfilter.BCFTools.BCFTools.subset_vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset_vcf</span></code></a>(outprefix[,&nbsp;bed,&nbsp;region,&nbsp;outdir,&nbsp;…])</td>
<td>Subset the vcf file using a BED file/region having the coordinates of the variants to exclude/include</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFfilter.BCFTools.BCFTools.filter">
<code class="descname">filter</code><span class="sig-paren">(</span><em>name</em>, <em>expression</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.BCFTools.BCFTools.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Run bcftools filter on a Vcf file</p>
<dl class="docutils">
<dt>name :str, Required</dt>
<dd>annotate FILTER column with &lt;str&gt;</dd>
<dt>expression :str, Required</dt>
<dd>exclude sites for which expression is true. i.e. ‘INFO/DP&gt;24304 | MQ&lt;34’</dd>
<dt>verbose <span class="classifier-delimiter">:</span> <span class="classifier">Boolean, optional</span></dt>
<dd>Increase verbosity</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Returns the path for the filtered file</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFfilter.BCFTools.BCFTools.filter_by_variant_type">
<code class="descname">filter_by_variant_type</code><span class="sig-paren">(</span><em>outprefix</em>, <em>v_type='snps'</em>, <em>compress=True</em>, <em>biallelic=False</em>, <em>action='select'</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.BCFTools.BCFTools.filter_by_variant_type" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>Method to filter a VCF file by variant type. For example, to extract only the SNPs</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>v_type</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required. Valid values are ‘snps’/’indels’,’mnps’,’other’,’both’. Default=snps</span></dt>
<dd><blockquote class="first">
<div><p>Extract/Filter (depending on the value of the ‘action’
argument) a certain variant type</p>
</div></blockquote>
<dl class="docutils">
<dt>compress <span class="classifier-delimiter">:</span> <span class="classifier">bool, Optional</span></dt>
<dd><p class="first last">If True then generate a vcf.gz file. Default=True</p>
</dd>
<dt>biallelic <span class="classifier-delimiter">:</span> <span class="classifier">bool, Optional</span></dt>
<dd><p class="first last">Select only biallelic variants. Default=False</p>
</dd>
</dl>
<p>action : str, Required. Valid values are ‘select’/’exclude’. Default=select
outprefix : str, Required. Prefix used for the output files
verbose : Boolean, optional</p>
<blockquote class="last">
<div><p>Increase verbosity</p>
</div></blockquote>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFfilter.BCFTools.BCFTools.select_variants">
<code class="descname">select_variants</code><span class="sig-paren">(</span><em>outprefix</em>, <em>uncalled=None</em>, <em>threads=1</em>, <em>verbose=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.BCFTools.BCFTools.select_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Run bcftools view to select only the variants (exclude the 0|0 genotypes)</p>
<p>outprefix : str, Required. Prefix used for the output file
uncalled : str, optional. Select/Exclude sites with an uncalled genotype.</p>
<blockquote>
<div>Possible values are: ‘exclude’, ‘include’</div></blockquote>
<dl class="docutils">
<dt>threads: int, optional</dt>
<dd>Number of output compression threads to use in addition to main thread. Default=0</dd>
<dt>verbose <span class="classifier-delimiter">:</span> <span class="classifier">Boolean, optional</span></dt>
<dd>Increase verbosity</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Returns the path for the VCF with only the variants</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFfilter.BCFTools.BCFTools.subset_vcf">
<code class="descname">subset_vcf</code><span class="sig-paren">(</span><em>outprefix</em>, <em>bed=None</em>, <em>region=None</em>, <em>outdir=None</em>, <em>create_index=False</em>, <em>verbose=None</em>, <em>action='exclude'</em>, <em>apply_filters=None</em>, <em>threads=1</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.BCFTools.BCFTools.subset_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset the vcf file using a BED file/region having the coordinates of the
variants to exclude/include</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>bed</strong> <span class="classifier-delimiter">:</span> <span class="classifier">string, Optional</span></dt>
<dd><p class="first last">BED file with coordinates to exclude/include</p>
</dd>
<dt><strong>region</strong> <span class="classifier-delimiter">:</span> <span class="classifier">string, Optional</span></dt>
<dd><p class="first last">String with region to consider: chr1, chr1:1000-1500, etc…</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, required</span></dt>
<dd><p class="first last">prefix for outputfiles</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">If provided, then put output files in this folder</p>
</dd>
<dt><strong>create_index</strong> <span class="classifier-delimiter">:</span> <span class="classifier">Boolean, optional</span></dt>
<dd><p class="first last">Generate a tabix index. Default=False</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">Boolean, optional</span></dt>
<dd><p class="first last">verbose</p>
</dd>
<dt><strong>action: str, optional</strong></dt>
<dd><p class="first last">Exclude or include variants from the bed file passed through the
bed option. Default= exclude</p>
</dd>
<dt><strong>apply_filters: str, optional</strong></dt>
<dd><p class="first last">Apply a filter string: i.e. “PASS,.”</p>
</dd>
<dt><strong>threads: int, optional</strong></dt>
<dd><p class="first last">Number of output compression threads to use in addition to main thread. Default=0</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Path to gzipped VCF file that will have the desired variants excluded/included</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-VCF.VCFfilter.GATK">
<span id="vcf-vcffilter-gatk-module"></span><h2>VCF.VCFfilter.GATK module<a class="headerlink" href="#module-VCF.VCFfilter.GATK" title="Permalink to this headline">¶</a></h2>
<p>Created on 27 Feb 2017</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="VCF.VCFfilter.GATK.GATK">
<em class="property">class </em><code class="descclassname">VCF.VCFfilter.GATK.</code><code class="descname">GATK</code><span class="sig-paren">(</span><em>vcf</em>, <em>reference</em>, <em>caller='UG'</em>, <em>bgzip_folder=None</em>, <em>tabix_folder=None</em>, <em>gatk_folder=None</em>, <em>tmp_dir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to filter a VCF file using different GATK components</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#VCF.VCFfilter.GATK.GATK.run_applyrecalibration" title="VCF.VCFfilter.GATK.GATK.run_applyrecalibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_applyrecalibration</span></code></a>(mode,&nbsp;recal_file,&nbsp;…)</td>
<td>Run GATK’s ApplyRecalibration on a VcfFilter object</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#VCF.VCFfilter.GATK.GATK.run_variantrecalibrator" title="VCF.VCFfilter.GATK.GATK.run_variantrecalibrator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_variantrecalibrator</span></code></a>(resources,&nbsp;mode[,&nbsp;…])</td>
<td>Run GATK’s VariantRecalibrator on a VcfFilter object</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="VCF.VCFfilter.GATK.GATK.run_applyrecalibration">
<code class="descname">run_applyrecalibration</code><span class="sig-paren">(</span><em>mode</em>, <em>recal_file</em>, <em>tranches_file</em>, <em>outprefix</em>, <em>ts_filter_level=99.0</em>, <em>num_threads=1</em>, <em>compress=True</em>, <em>verbose=None</em>, <em>log_file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK.run_applyrecalibration" title="Permalink to this definition">¶</a></dt>
<dd><p>Run GATK’s ApplyRecalibration on a VcfFilter object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>mode</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">Recalibration mode to employ (SNP|INDEL)</p>
</dd>
<dt><strong>recal_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">The input recal file used by ApplyRecalibration</p>
</dd>
<dt><strong>tranches_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">The input tranches file describing where to cut the data</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">Prefix used for the output</p>
</dd>
<dt><strong>ts_filter_level</strong> <span class="classifier-delimiter">:</span> <span class="classifier">float, Optional</span></dt>
<dd><p class="first last">The truth sensitivity level at which to start filtering. Default=99.0</p>
</dd>
<dt><strong>num_threads</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, Optional</span></dt>
<dd><p class="first last">Number of data threads to allocate to this analysis. Default=1</p>
</dd>
<dt><strong>compress</strong> <span class="classifier-delimiter">:</span> <span class="classifier">boolean, Default= True</span></dt>
<dd><p class="first last">Compress the recalibrated VCF</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, Optional</span></dt>
<dd><p class="first last">Increase verbosity</p>
</dd>
<dt><strong>log_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">Path to file that will used for logging the GATK stderr and stdout</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="VCF.VCFfilter.GATK.GATK.run_variantrecalibrator">
<code class="descname">run_variantrecalibrator</code><span class="sig-paren">(</span><em>resources</em>, <em>mode</em>, <em>max_gaussians=None</em>, <em>intervals=None</em>, <em>annotations=None</em>, <em>tranches=None</em>, <em>outprefix='recalibrate'</em>, <em>verbose=None</em>, <em>log_file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#VCF.VCFfilter.GATK.GATK.run_variantrecalibrator" title="Permalink to this definition">¶</a></dt>
<dd><p>Run GATK’s VariantRecalibrator on a VcfFilter object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>resources</strong> <span class="classifier-delimiter">:</span> <span class="classifier">JSON file with resources to add using the -resources option, Required</span></dt>
<dd></dd>
<dt><strong>mode</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Required</span></dt>
<dd><p class="first last">Recalibration mode to employ (SNP|INDEL)</p>
</dd>
<dt><strong>intervals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">chr1:1-1000, Optional</span></dt>
<dd><p class="first last">One or more genomic intervals over which to operate</p>
</dd>
<dt><strong>max_gaussians</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, Optional</span></dt>
<dd><p class="first last">Max number of Gaussians for the positive model</p>
</dd>
<dt><strong>annotations</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list, Optional</span></dt>
<dd><p class="first last">List of annotations to be used. Default=[‘DP’, ‘QD’, ‘FS’, ‘SOR’,
‘MQ’, ‘MQRankSum’, ‘ReadPosRankSum’, ‘InbreedingCoeff’]</p>
</dd>
<dt><strong>tranches</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list, Optional</span></dt>
<dd><p class="first last">Each element in the list will correspond to the  levels of truth
sensitivity at which to slice the data. (in percent, that is 1.0
for 1 percent). Default=[100.0,99.9,99.0,90.0]</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">out prefix used for -recalFile, -tranchesFile, -rscriptFile.
Default= recalibrate</p>
</dd>
<dt><strong>verbose</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool, Optional</span></dt>
<dd><p class="first last">Increase verbosity</p>
</dd>
<dt><strong>log_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, Optional</span></dt>
<dd><p class="first last">Path to file that will used for logging the GATK stderr and stdout</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>Dictionary with location of tranches and recal files</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-VCF.VCFfilter">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-VCF.VCFfilter" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="VCF.html">VCF package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="VCF.html#subpackages">Subpackages</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfNormalize">VCF.VcfNormalize module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfQC">VCF.VcfQC module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF.VcfUtils">VCF.VcfUtils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="VCF.html#module-VCF">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage/installation.html">Installation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="VCF.html">VCF package</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/VCF.VCFfilter.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>