/* 
 * Workflow to filter a VCF
 * This workflow relies on Nextflow (see https://www.nextflow.io/tags/workflow.html)
 *
 * @author
 * Ernesto Lowy <ernesto.lowy@gmail.com>
 *
 */

// params defaults
params.help = false
params.threads = 1
params.queue = 'production-rh7'

//print usage
if (params.help) {
    log.info ''
    log.info 'Pipeline to apply a certain fitted Logistic Regression model to filter a certain VCF file'
    log.info '-----------------------------------------------------------------------------------------'
    log.info ''
    log.info 'Usage: '
    log.info '    nextflow MLfilter_applymodel.nf --vcf VCF --model --cutoff 0.95 --threads 5 --vt snps'
    log.info ''
    log.info 'Options:'
    log.info '	--help	Show this message and exit.'
    log.info '	--vcf VCF    Path to the VCF file that will be filtered.'
    log.info '  --model FILE Path to serialized ML fitted model.'
    log.info '  --cutoff FLOAT Cutoff value used in the filtering.'
    log.info '  --annotations ANNOTATION_STRING String containing the annotations to filter, for example:'
    log.info '  --vt  VARIANT_TYPE   Type of variant to filter. Poss1ible values are 'snps'/'indels'.'
    log.info '  %CHROM\t%POS\t%INFO/DP\t%INFO/RPB\t%INFO/MQB\t%INFO/BQB\t%INFO/MQSB\t%INFO/SGB\t%INFO/MQ0F\t%INFO/ICB\t%INFO/HOB\t%INFO/MQ\n.'
    log.info '  --threads INT Number of threads used in the different BCFTools processes. Default=1.'
    log.info ''
    exit 1
}


chrList=['chr20']

chrChannel=Channel.from( chrList )


//Apply a fitted model obtained after running MLfilter_trainmodel.nf

process get_variant_annotations {
	/*
	Process to get the variant annotations for the selected ${params.vt} from the unfiltered VCF file
	*/

	memory '2 GB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	input:
	val chr from chrChannel

	output:
	file 'unfilt_annotations.vt.tsv.gz' into unfilt_annotations
	val chr into chr

	"""
	bcftools view -c1 -r ${chr} -v ${params.vt} ${params.vcf} -o out.onlyvariants.vt.vcf.gz -Oz --threads ${params.threads}
	tabix out.onlyvariants.vt.vcf.gz
	bcftools query -H -r ${chr} -f '${params.annotations}' out.onlyvariants.vt.vcf.gz | bgzip -c > unfilt_annotations.vt.tsv.gz
	"""
}

process apply_model {
	/*
	Process to read-in the serialized ML model created by running MLfilter_trainmodel.nf
	and to apply this model on the unfiltered VCF
	*/

	memory '2 GB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	input:
	file unfilt_annotations

	output:
	file 'predictions.tsv' into predictions

	"""
	#!/usr/bin/env python

	from VCF.VCFfilter.MLclassifier import MLclassifier

	ML_obj=MLclassifier(fitted_model = '${params.model}')

	ML_obj.predict(outprefix="predictions", annotation_f='${unfilt_annotations}', cutoff=${params.cutoff})
	"""
}

process compress_predictions {
	/*
	Process to compress and index the 'predictions.tsv' file generated by process 'apply_model'
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	input:
	file predictions

	output:
	file 'predictions.tsv.gz' into predictions_table
	file 'predictions.tsv.gz.tbi' into predictions_table_tabix

	"""
	bgzip -c ${predictions} > 'predictions.tsv.gz'
	tabix -f -s1 -b2 -e2 'predictions.tsv.gz'
	"""
}

process get_header {
	/*
	Process to get the header of the unfiltered VCF
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	output:
	file 'header.txt' into header

	"""
	bcftools view -h ${params.vcf} > header.txt
	"""
}

process modify_header {
	/*
	Process to modify the header of the unfiltered VCF
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	input:
	file header

	output:
	file 'newheader.txt' into newheader

	"""
	#!/usr/bin/env python

	from VCF.VcfUtils import VcfUtils

	vcf_object=VcfUtils(vcf='${params.vcf}')

	vcf_object.add_to_header(header_f='${header}', outfilename='newheader1.txt',
                          	 line_ann='##FILTER=<ID=MLFILT,Description="Binary classifier filter">')
	vcf_object.add_to_header(header_f='newheader1.txt', outfilename='newheader.txt',
                                 line_ann='##INFO=<ID=prob_TP,Number=1,Type=Float,Description="Probability of being a True positive">')
	"""
}

process splitVCF {
	/*
	This process will select a single chromosome from the VCF
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus "${params.threads}"

	input:
	val chr

	output:
	val chr into chr_1
	file "unfilt.${chr}.vcf.gz" into unfilt_vcf_chr
	
	"""
	bcftools view -r ${chr} ${params.vcf} -o unfilt.${chr}.vcf.gz --threads ${params.threads} -Oz
	"""
}

process replace_header {
	/*
	Process to replace header in the unfiltered VCF
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus 1

	input:
	file newheader
	file unfilt_vcf_chr

	output:
	file 'unfilt_reheaded.vcf.gz' into unfilt_vcf_chr_reheaded

	"""
	bcftools reheader -h ${newheader} -o 'unfilt_reheaded.vcf.gz' ${unfilt_vcf_chr}
	"""
}

process reannotate_vcf {
	/*
	Process to reannotate the unfiltered VCF with the information generated after applying the classifier
	*/

	memory '500 MB'
        executor 'local'
        queue "${params.queue}"
        cpus "${params.threads}"

	publishDir "results_${chr_1}", mode: 'copy', overwrite: true

	input:
	file unfilt_vcf_chr_reheaded
	file predictions_table
	file predictions_table_tabix
	val chr_1

	output:
	file 'filt.vcf.gz' into filt_vcf

	"""
	bcftools annotate -a ${predictions_table} ${unfilt_vcf_chr_reheaded} -c CHROM,POS,FILTER,prob_TP -o filt.vcf.gz --threads ${params.threads} -Oz
	"""
}
