
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>BamQC &#8212; igsr_analysis 0.1.1 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="bamqc">
<h1>BamQC<a class="headerlink" href="#bamqc" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<span class="target" id="module-BamQC.BamQC"></span><p>Created on 12 Oct 2016</p>
<p>&#64;author: ernesto</p>
<dl class="class">
<dt id="BamQC.BamQC.BamQC">
<em class="property">class </em><code class="descclassname">BamQC.BamQC.</code><code class="descname">BamQC</code><span class="sig-paren">(</span><em>bam</em>, <em>samtools_folder=None</em>, <em>java_folder=None</em>, <em>picard_folder=None</em>, <em>chk_indel_folder=None</em>, <em>verifybamid_folder=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to do the quality assessment on a BAM format file</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.aggregate_stats" title="BamQC.BamQC.BamQC.aggregate_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aggregate_stats</span></code></a>(cov_list)</td>
<td>Used to calculate aggregated stats on a list of SDepth objects</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.get_contigs" title="BamQC.BamQC.BamQC.get_contigs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_contigs</span></code></a>()</td>
<td>Get all contigs from this BAM</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.get_simple_stats" title="BamQC.BamQC.BamQC.get_simple_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_simple_stats</span></code></a>()</td>
<td>Get a dict with stats on the BAM file as calculated by samtools flagstat</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.list_of_readgroups" title="BamQC.BamQC.BamQC.list_of_readgroups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">list_of_readgroups</span></code></a>()</td>
<td>Get the Read Groups extracted from the header of the BAM file</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.list_of_samples" title="BamQC.BamQC.BamQC.list_of_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">list_of_samples</span></code></a>()</td>
<td>Get the samples names from the header of the BAM file</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.run_CollectHsMetrics" title="BamQC.BamQC.BamQC.run_CollectHsMetrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_CollectHsMetrics</span></code></a>(baits_file[,&nbsp;outfile,&nbsp;…])</td>
<td>Run Picard’s CollectHsMetrics on a Exome sequencing BAM file</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.run_CollectWgsMetrics" title="BamQC.BamQC.BamQC.run_CollectWgsMetrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_CollectWgsMetrics</span></code></a>(reference[,&nbsp;outfile,&nbsp;…])</td>
<td>Run Picard’s CollectWgsMetrics on a WGS BAM file</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.run_chk_indel_rg" title="BamQC.BamQC.BamQC.run_chk_indel_rg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_chk_indel_rg</span></code></a>([outfile])</td>
<td>Run Heng Li’s chk_indel_rg on a BAM file</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.run_samtools_depth" title="BamQC.BamQC.BamQC.run_samtools_depth"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_samtools_depth</span></code></a>(chros)</td>
<td>Calculate several coverage metrics on a whole genome sequencing BAM file using ‘samtools depth’</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.BamQC.run_verifybamid" title="BamQC.BamQC.BamQC.run_verifybamid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_verifybamid</span></code></a>(genotype_file,&nbsp;outprefix[,&nbsp;…])</td>
<td>Run VerifyBAMID to check for sample swap or contamination issues</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="BamQC.BamQC.BamQC.aggregate_stats">
<code class="descname">aggregate_stats</code><span class="sig-paren">(</span><em>cov_list</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.aggregate_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to calculate aggregated stats on a list of SDepth objects</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>cov_list</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd><p class="first last">List containing the SDepth objects for which the stats will be aggregated.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>A SDepth object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.get_contigs">
<code class="descname">get_contigs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.get_contigs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all contigs from this BAM</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>None</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dict</strong></dt>
<dd><p class="first">A dictionary containing the following information:</p>
<blockquote class="last">
<div><p>{‘contig Name’: length (in bp)}</p>
</div></blockquote>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.get_simple_stats">
<code class="descname">get_simple_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.get_simple_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a dict with stats on the BAM file as calculated by samtools flagstat</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>None</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dict</strong></dt>
<dd><p class="first">A dictionary containing the following information:
{</p>
<blockquote class="last">
<div><p>“total_no_reads”: int
“no_duplicates”: int
“total_no_mapped”: int
“no_properly_paired”:  int
}</p>
</div></blockquote>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.list_of_readgroups">
<code class="descname">list_of_readgroups</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.list_of_readgroups" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the Read Groups extracted from the header of the BAM file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>None</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>list</strong></dt>
<dd><p class="first last">List composed of the read groups</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.list_of_samples">
<code class="descname">list_of_samples</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.list_of_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the samples names from the header of the BAM file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>None</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>list</strong></dt>
<dd><p class="first last">List with the sample names</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.run_CollectHsMetrics">
<code class="descname">run_CollectHsMetrics</code><span class="sig-paren">(</span><em>baits_file</em>, <em>outfile=None</em>, <em>cov_cap=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.run_CollectHsMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Run Picard’s CollectHsMetrics on a Exome sequencing BAM file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>baits_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">Path to the file containing the Exome baits.</p>
</dd>
<dt><strong>outfile</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">If provided, then create a file with the output of this program</p>
</dd>
<dt><strong>cov_cap</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">Picard’s Coverage Cap parameter. Treat positions with coverage
exceeding this value as if they had coverage at this value.
Default value: 250.</p>
</dd>
<dt><strong>Returns</strong></dt>
<dd></dd>
<dt><strong>——</strong></dt>
<dd></dd>
<dt><strong>A CMetrics object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.run_CollectWgsMetrics">
<code class="descname">run_CollectWgsMetrics</code><span class="sig-paren">(</span><em>reference</em>, <em>outfile=None</em>, <em>cov_cap=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.run_CollectWgsMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Run Picard’s CollectWgsMetrics on a WGS BAM file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>reference</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">Fasta file used as the genome reference</p>
</dd>
<dt><strong>outfile</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">If provided, then create a file with the output of this program</p>
</dd>
<dt><strong>cov_cap</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd><p class="first last">Picard’s Coverage Cap parameter. Treat positions with coverage
exceeding this value as if they had coverage at this value.
Default value: 250.</p>
</dd>
<dt><strong>Returns</strong></dt>
<dd></dd>
<dt><strong>——</strong></dt>
<dd></dd>
<dt><strong>A CMetrics object</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.run_chk_indel_rg">
<code class="descname">run_chk_indel_rg</code><span class="sig-paren">(</span><em>outfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.run_chk_indel_rg" title="Permalink to this definition">¶</a></dt>
<dd><p>Run Heng Li’s chk_indel_rg on a BAM file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>outfile</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">If provided, then create a file with the output of this program</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>list</strong></dt>
<dd><p class="first last">A list of Chk_indel objects</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.run_samtools_depth">
<code class="descname">run_samtools_depth</code><span class="sig-paren">(</span><em>chros</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.run_samtools_depth" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate several coverage metrics on a whole genome sequencing BAM
file using ‘samtools depth’</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>chros</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list or string</span></dt>
<dd><p class="first last">List of contigs or just a single contig used for calculating the coverage</p>
</dd>
<dt><strong>Returns</strong></dt>
<dd></dd>
<dt><strong>——</strong></dt>
<dd></dd>
<dt><strong>List of SDepth objects</strong></dt>
<dd></dd>
<dt><strong>This method runs samtools depth on a BAM file and will calculate the following metrics:</strong></dt>
<dd><ul class="first last simple">
<li>Number of Bases mapped: This is the number of bases having at least one read mapped</li>
<li>Sum of depths of coverage: This is the sum of all the depths in each of the Bases mapped</li>
<li>Breadth of coverage: This is the result of dividing bases_mapped/length(contig)
(i.e. what portion of the contig has reads mapped)</li>
<li>Depth of coverage: This is the result of dividing sum_of_depths/length(contig)</li>
</ul>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.BamQC.run_verifybamid">
<code class="descname">run_verifybamid</code><span class="sig-paren">(</span><em>genotype_file</em>, <em>outprefix</em>, <em>outdir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.BamQC.run_verifybamid" title="Permalink to this definition">¶</a></dt>
<dd><p>Run VerifyBAMID to check for sample swap or contamination issues</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>genotype_file</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">vcf file with chip genotypes to use</p>
</dd>
<dt><strong>outprefix</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">prefix for outputfiles</p>
</dd>
<dt><strong>outdir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd><p class="first last">If provided, then put output files in this folder</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>list</strong></dt>
<dd><p class="first last">A list with the paths to the output files generated by VerifyBAMID</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="BamQC.BamQC.CMetrics">
<em class="property">class </em><code class="descclassname">BamQC.BamQC.</code><code class="descname">CMetrics</code><span class="sig-paren">(</span><em>metrics</em>, <em>cov_data</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.CMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store coverage information on the metrics calculated by Picard’s
CollectHsMetrics/CollectWgsMetrics on an Exome or WGS BAM file</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.CMetrics.create_cov_barplot" title="BamQC.BamQC.CMetrics.create_cov_barplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_cov_barplot</span></code></a>(filename[,&nbsp;xlim,&nbsp;ylim])</td>
<td>This method will create a Barplot using the different coverage values counts calculated by Picard’s CollectHsMetrics or CollectWgsMetrics</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#BamQC.BamQC.CMetrics.print_report" title="BamQC.BamQC.CMetrics.print_report"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_report</span></code></a>([filename])</td>
<td>Used to print a text report of data in the object</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="BamQC.BamQC.CMetrics.create_cov_barplot">
<code class="descname">create_cov_barplot</code><span class="sig-paren">(</span><em>filename</em>, <em>xlim=None</em>, <em>ylim=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.CMetrics.create_cov_barplot" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will create a Barplot using the different coverage
values counts calculated by Picard’s
CollectHsMetrics or CollectWgsMetrics</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename</span></dt>
<dd><p class="first last">PDF file to write the plot.</p>
</dd>
<dt><strong>xlim</strong> <span class="classifier-delimiter">:</span> <span class="classifier">tuple, optional</span></dt>
<dd><p class="first last">Set the X-axis limit</p>
</dd>
<dt><strong>ylim</strong> <span class="classifier-delimiter">:</span> <span class="classifier">tuple, optional</span></dt>
<dd><p class="first last">Set the Y-axis limit</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BamQC.BamQC.CMetrics.print_report">
<code class="descname">print_report</code><span class="sig-paren">(</span><em>filename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.CMetrics.print_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to print a text report of data in the object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">Filename to write the report. The default is STDOUT.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="BamQC.BamQC.Chk_indel">
<em class="property">class </em><code class="descclassname">BamQC.BamQC.</code><code class="descname">Chk_indel</code><span class="sig-paren">(</span><em>RG</em>, <em>ins_in_short_homopolymer</em>, <em>del_in_short</em>, <em>ins_in_long</em>, <em>del_in_long</em>, <em>outcome=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.Chk_indel" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store information on the ratio of short insertion and deletion
calculated by runnint Heng Li’s chk_indel_rg</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.Chk_indel.calc_ratio" title="BamQC.BamQC.Chk_indel.calc_ratio"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_ratio</span></code></a>()</td>
<td>Method to calc ratio ins-in-short-homopolymer/del-in-short and check if it is &gt; 5</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="BamQC.BamQC.Chk_indel.calc_ratio">
<code class="descname">calc_ratio</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.Chk_indel.calc_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calc ratio ins-in-short-homopolymer/del-in-short and check if it is &gt; 5</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>str</strong></dt>
<dd><p class="first last">It returns PASS/FAILED depending on the outcome of the test</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="BamQC.BamQC.SDepth">
<em class="property">class </em><code class="descclassname">BamQC.BamQC.</code><code class="descname">SDepth</code><span class="sig-paren">(</span><em>contig=None</em>, <em>mapped=None</em>, <em>breadth=None</em>, <em>depth=None</em>, <em>length=None</em>, <em>sum_of_depths=None</em>, <em>max=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.SDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to store coverage metrics on a Whole Genome Sequencing BAM file
calculated using SAMtools depth</p>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#BamQC.BamQC.SDepth.print_report" title="BamQC.BamQC.SDepth.print_report"><code class="xref py py-obj docutils literal notranslate"><span class="pre">print_report</span></code></a>([filename])</td>
<td>Used to print a text report of data in the object</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="BamQC.BamQC.SDepth.print_report">
<code class="descname">print_report</code><span class="sig-paren">(</span><em>filename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BamQC.BamQC.SDepth.print_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to print a text report of data in the object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>filename</strong> <span class="classifier-delimiter">:</span> <span class="classifier">filename, optional</span></dt>
<dd><p class="first last">Filename used to write the report. The default is STDOUT.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">igsr_analysis</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Auto-documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../VCF/VCF.html">VCF</a></li>
<li class="toctree-l2"><a class="reference internal" href="../BEDTools/BEDTools.html">BEDTools</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">BamQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../VariantCalling/VariantCalling.html">VariantCalling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Utils/Utils.html">Utils</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Auto-documentation</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, ernesto lowy.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../_sources/code/BamQC/BamQC.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>